bot/cogs/chain_cog.py

```diff
@@ -0,0 +1,104 @@
+import logging
+
+import discord
+import discord.ext.commands as commands
+from copy import copy
+
+import bot.extensions as ext
+
+log = logging.getLogger(__name__)
+
+class ChainCog(commands.Cog):
+    """This cog handles the chaining of commands"""
+
+    def __init__(self, bot):
+        self.bot = bot
+        self.resultqueue = []
+
+    @ext.command()
+    @ext.chainable_output(False)
+    @ext.long_help('Chains multiple chainable commands together')
+    @ext.short_help('Chains commands')
+    @ext.example('chain command1 command2 hello')
+    async def chain(self, ctx, *, text: str):
+        prefix = str(await self.bot.current_prefix(ctx))
+        # Make a fake ctx to intercept values passed to send
+        fakectx = copy(ctx)
+        fakectx.send = self.fakesend
+        # The first command doesnt have to have a chainable output so processed later
+        try:
+            command1 = text[text.index(prefix)+1:text.index(' ')]
+        except ValueError:
+            await ctx.send("no valid first command")
+            return
+        remainder = text[text.index(' ')+1:]
+        # Get output of the chainable output commands
+        while(True):
+            currentcommandindex = remainder.rfind(prefix)
+            if currentcommandindex == -1:
+                # If no more commands, break
+                break
+            currentcommand = self.find_command(self.bot, remainder[currentcommandindex+1:remainder.index(' ',currentcommandindex)])
+            if currentcommand is None:
+                # if the command isnt valid, stop execution
+                await ctx.send(remainder[currentcommandindex + 1:remainder.index(' ', currentcommandindex)]+' is not a valid command')
+                return
+            if currentcommand.chainable_output:
+                argname = str(next(reversed(currentcommand.params.values())))
+                buffer = argname.find('=')
+                if buffer != -1:
+                    argname = argname[:buffer]
+                # Run the command passing the fake ctx
+                await eval('currentcommand(' + argname + '= \''+remainder[remainder.index(' ', currentcommandindex)+1:]+'\', ctx=fakectx)')
```

Comment:
https://discordpy.readthedocs.io/en/latest/ext/commands/api.html?highlight=invoke#discord.ext.commands.Context.invoke

Reply:
this is the big one, we CAN NOT under any circumstances use eval